### ISSUE TITLE
Kube2iam daemonset not applying specified role to deployment, falling back to default role

### ISSUE THREAD

**Issue by saidmasoud** [jtblin/kube2iam#175]
I am running kube2iam in a kops cluster, and trying to assign an IAM role to a deployment. I have added the appropriate trust realtionship to the role. However, I cannot get the deployment to assume the correct role; it is instead assuming the default node role.

Kube2iam spec (RBAC has been applied and daemon successfully runs in cluster):
```yaml
apiVersion: apps/v1
kind: DaemonSet
metadata:
  namespace: kube-system
  name: kube2iam
  labels:
    app: kube2iam
spec:
  selector:
    matchLabels:
      name: kube2iam
  template:
    metadata:
      labels:
        name: kube2iam
    spec:
      serviceAccountName: kube2iam
      hostNetwork: true
      containers:
        - image: jtblin/kube2iam:latest
          name: kube2iam
          args:
            - "--auto-discover-base-arn"
            - "--auto-discover-default-role=true"
            - "--node=$(NODE_NAME)"
            - "--debug"
            - "--use-regional-sts-endpoint"
          env:
            - name: NODE_NAME
              valueFrom:
                fieldRef:
                  fieldPath: spec.nodeName
            - name: AWS_REGION
              value: "ca-central-1"
          ports:
            - containerPort: 8181
              hostPort: 8181
              name: http
```

Deployment spec:
```yaml
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: external-dns
spec:
  strategy:
    type: Recreate
  template:
    metadata:
      annotations:
        iam.amazonaws.com/role: external-dns
      labels:
        app: external-dns
    spec:
      serviceAccountName: external-dns
      containers:
      - name: external-dns
        image: registry.opensource.zalan.do/teapot/external-dns:v0.5.6
        args:
        - --source=service
        - --source=ingress
        - --domain-filter=sub.domain.com # change this before applying
        - --provider=aws
        - --policy=sync
        - --aws-zone-type=public # only look at public hosted zones (valid values are public, private or no value for both)
        - --registry=txt
        - --txt-owner-id=my-identifier
```

Deployment logs:
```
time="2018-10-24T13:20:38Z" level=error msg="AccessDenied: User: arn:aws:sts::<account-id>:assumed-role/nodes.k8s.sub.domain.com/i-<instance-id> is not authorized to perform: route53:ListHostedZones\n\tstatus code: 403, request id: <request-id>"
```

**Comment by stafot** [comment]
Facing exact the same problem on EKS. 
Logs from deployment.
 `[2018-11-06T09:01:57,087][WARN ][logstash.outputs.elasticsearch] Attempted to resurrect connection to dead ES instance, but got an error. {:url=>"http://es-logging-staging.hourlie.com:80/", :error_type=>LogStash::Outputs::ElasticSearch::HttpClient::Pool::BadResponseCodeError, :error=>"Got response code '403' contacting Elasticsearch at URL 'http://<my_aws_es_endpoint/'"}`
Logs from `kube2iam`:
`time="2018-11-06T08:49:30Z" level=error msg="Error assuming role AccessDenied: Access denied\n\tstatus code: 403, request id: e01c5506-e1a0-11e8-949f-3717b9a3b42b" ns.name=default pod.iam.role="arn:aws:iam::<my_account_id>:role/<my_worker_role_name>" req.method=GET req.path=/latest/meta-data/iam/security-credentials/eks_worker req.remote=10.56.0.163`

Apart from the policies that needed by `kube2iam` there is also setup properly the role on AWS Elasticsearch service access policy.

**Comment by metacyclic** [comment]
I have faced similar problems, and it is always because the node instance role of the eks worker node group was not added in the trust relationship of the role that was being assumed.  The role you are trying to assume should have something like this as a trust relationship

```
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Principal": {
        "AWS": "arn:aws:iam::<account_id>:role/<name of eks clsuter>-NodeInstanceRole-<some identifier>"
      },
      "Action": "sts:AssumeRole"
    }
  ]
}
```

Node instance role can be found in the `Resources` tab of your CloudFormation stack

**Comment by metacyclic** [comment]
Hmm, the error you are seeing seems to indicate that it isn't even trying to assume the role.

Try adding
```

**Comment by metacyclic** [comment]
@stafot - that looks right.  Perhaps terminate and redeploy your pod/deployment?  Have you redeployed since you started running kube2iam?

**Comment by saidmasoud** [comment]
@metacyclic that did it, thanks so much! Here's my final Daemonset definition:

```
apiVersion: apps/v1
kind: DaemonSet
metadata:
  namespace: kube-system
  name: kube2iam
  labels:
    app: kube2iam
spec:
  selector:
    matchLabels:
      name: kube2iam
  template:
    metadata:
      labels:
        name: kube2iam
    spec:
      serviceAccountName: kube2iam
      hostNetwork: true
      containers:
        - image: jtblin/kube2iam:latest
          name: kube2iam
          args:
            - "--auto-discover-default-role=true"
            - "--node=$(NODE_NAME)"
            - "--debug"
            - "--use-regional-sts-endpoint"
            - "--base-role-arn=arn:aws:iam::123456789012:role/"
            - "--iptables=true"
            - "--host-ip=$(HOST_IP)"
            - "--host-interface=cbr0"
          env:
            - name: HOST_IP
              valueFrom:
                fieldRef:
                  fieldPath: status.podIP
            - name: NODE_NAME
              valueFrom:
                fieldRef:
                  fieldPath: spec.nodeName
            - name: AWS_REGION
              value: "ca-central-1"
          ports:
            - containerPort: 8181
              hostPort: 8181
              name: http
          securityContext:
            privileged: true
```

**Comment by stafot** [comment]
@metacyclic No I didn't delete and recreate deployment. Just replaced. But replace does change the pod. So It should work. Anyway I will try this also.

**Comment by metacyclic** [comment]
@stafot - this part ( seems to be part of your trust relationship )
```
{
  "Sid": "",
  "Effect": "Allow",
  "Principal": {
    "AWS": "arn:aws:iam::<my-account-id>:role/<my_node_role_name>"
    },
    "Action": "sts:AssumeRole"
}
```
should be in the `trust relationships` section of your role ( tab to the right of `permissions`), and you should keep `eni+` as you have it already.  Your error points to a role setup problem `Error assuming role AccessDenied: Access denied`, and seems like kube2iam is working as expected.

Good luck!

**Comment by jicowan** [comment]
@saidmasoud I am experiencing a similar issue.  I assigned the fluentd daemonset an IAM role and updated access policy in AWS ES to use an resource based policy with that role.  When I look at the fluentd logs, I see 403 errors [forbidden].  It's not until I change the access policy to allow anonymous connections that fluentd is able to post logs to ES. When I exec into a fluentd pod and curl the EC2 metadata endpoint, I can see that the pod is using the fallback role (default) and the role I've assigned to it in the annotation.  It seems like fluentd is not using either of them when posting to ES.

**Comment by KernCheh** [comment]
Just for the record FWIW:
I've encountered errors in the format of
```
Error assuming role AccessDenied: Access denied\n\tstatus code: 403, request id: 
```
A simple access denied without any additional information.

As it turns out, this is expected behavior, it's likely because the pod does not have the correct role.

For my case, it was because I mixed up deployment annotations and pod annotations, thereby putting the role in the deployment annotation.
